### PR TITLE
Refactor update_icons

### PR DESCRIPTION
**Pull request by guidezpl**
## Description

Some preliminary refactoring needed to implement adaptive icons. 

## Related Issues

Enables https://github.com/flutter/flutter/issues/68016

## Tests

No tests needed, since the script produces the exact same material icons.dart file.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/tools/update_icons.dart [hunk 1]
```diff
@@ -74,7 +74,7 @@ const Map<String, String> _identifierRewrites = <String, String>{
   'class': 'class_',
 };
 
-const Set<String> _mirroredIcons = <String>{
+const Set<String> _mirroredInRTL = <String>{
```
Comment (JoseAlba) [inline dev/tools/update_icons.dart:77]: nit: is mirrored in rtl icons the same thing as ltr icons. I am not too sure if mirrordinRTL is not the same thing as LTR icons.

File: dev/tools/update_icons.dart [hunk 2]
```diff
@@ -262,52 +268,76 @@ Error: New codepoints file does not contain all the existing codepoints.\n
...
-    webFontKey = webFontKey.replaceAll('_rounded', '');
+enum IconStyle {
+  regular,
+  outlined,
+  rounded,
+  sharp,
+}
+
+extension IconStyleSuffix on IconStyle {
+  // The suffix for the 'material-icons' HTML class
```
Comment (JoseAlba) [inline dev/tools/update_icons.dart:279]: nit: add period at the end of sentence.

File: dev/tools/update_icons.dart [hunk 3]
```diff
@@ -262,52 +268,76 @@ Error: New codepoints file does not contain all the existing codepoints.\n
...
+  String suffix() {
+    switch (this) {
```
Comment (JoseAlba) [inline dev/tools/update_icons.dart:281]: this switch statement doesn't have a case for iconstyle.regular

File: dev/tools/update_icons.dart [hunk 4]
```diff
@@ -262,52 +268,76 @@ Error: New codepoints file does not contain all the existing codepoints.\n
...
+      default: return '';
+    }
   }
-  if (webFontKey.endsWith('_sharp')) {
-    styleSuffix = '-sharp';
-    webFontKey = webFontKey.replaceAll('_sharp', '');
+}
+
+class _Icon {
+  // Parse tokenPair (e.g. {"6_ft_apart_outlined": "e004"})
```
Comment (JoseAlba) [inline dev/tools/update_icons.dart:291]: period at the end of sentence.

**Comment by willlarche** [comment]
What's the plan to get from here to adaptive?

File: dev/tools/update_icons.dart (hunk 3 above)
Comment (JoseAlba) [reply dev/tools/update_icons.dart:281]: My assumption is regular has no prefix that is why it defaulted. I wonder in this case if it should still have its own case and it returning an empty string. I am not too sure about this. It seems like the way you implemented it looks cleaner.

**Review by JoseAlba — APPROVED** [review]
```
██╗      ██████╗ ████████╗███╗   ███╗
██║     ██╔════╝ ╚══██╔══╝████╗ ████║
██║     ██║  ███╗   ██║   ██╔████╔██║
██║     ██║   ██║   ██║   ██║╚██╔╝██║
███████╗╚██████╔╝   ██║   ██║ ╚═╝ ██║
╚══════╝ ╚═════╝    ╚═╝   ╚═╝     ╚═╝
```

File: dev/tools/update_icons.dart (hunk 1 above)
Comment (guidezpl) [reply dev/tools/update_icons.dart:77]: > What's the plan to get from here to adaptive?

A private class to make Icons.adaptive.X possible, and an update to the script to auto-generated that private class's getters based on icons that include "ios" in the name

**Comment by guidezpl** [comment]
> What's the plan to get from here to adaptive?

A private class to make Icons.adaptive.X possible, and an update to the script to auto-generate that private class' getters based on icons that include "ios" in the name

**Review by rami-a — APPROVED** [review]
LGTM

**Comment by willlarche** [comment]
> > What's the plan to get from here to adaptive?
> 
> A private class to make `Icons.adaptive.X` possible, and an update to the script to auto-generate that private class' getters based on icons that include "ios" in the name

Do all the adaptive icons have '_ios'?

File: dev/tools/update_icons.dart (hunk 3 above)
Comment (guidezpl) [reply dev/tools/update_icons.dart:281]: I would do it, but you end up having to do either of these to make the analyser happy, neither of which feel right

```
    switch (this) {
      case IconStyle.regular: return '';
      case IconStyle.outlined: return '-outlined';
      case IconStyle.rounded: return '-round';
      case IconStyle.sharp: return '-sharp';
    }
    return '';
```

```
    switch (this) {
      case IconStyle.regular: return '';
      case IconStyle.outlined: return '-outlined';
      case IconStyle.rounded: return '-round';
      case IconStyle.sharp: return '-sharp';
      default: return '';
    }
```

**Comment by guidezpl** [comment]
> > > What's the plan to get from here to adaptive?
> > 
> > 
> > A private class to make `Icons.adaptive.X` possible, and an update to the script to auto-generate that private class' getters based on icons that include "ios" in the name
> 
> Do all the adaptive icons have '_ios'?

Yes, except a special case for more (more_vert, more_horiz)

File: dev/tools/update_icons.dart (hunk 1 above)
Comment (guidezpl) [reply dev/tools/update_icons.dart:77]: this means these icons should always match text direction, thus be mirrored vertically when text direction is RTL

Comment (willlarche) [reply dev/tools/update_icons.dart:77]: Suggestion of 'iconsMirroredWhenRTL' if that's clearer.

Comment (guidezpl) [reply dev/tools/update_icons.dart:77]: Done

**Comment by willlarche** [comment]
LGTM

**Pull request merged**